include/barrier.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2022 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public interface for barriers.
+ */
+#ifndef ZEPHYR_INCLUDE_BARRIER_H_
+#define ZEPHYR_INCLUDE_BARRIER_H_
+
+#ifndef _ASMLANGUAGE
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup barrier_apis Barriers API
+ * @ingroup kernel_apis
+ * @{
+ */
+
+#if !defined(nop) || defined(__DOXYGEN__)
+/**
+ * @brief nop
+ *
+ * NOP instruction.
+ */
+#define nop()		__asm__ volatile ("nop")
+#endif /* !nop */
+
+#if !defined(isb) || defined(__DOXYGEN__)
```

Comment:
Not personally a fan of this pattern.  I know the intent is to allow the arch layer to override the definition, but what it really means is that if you aren't careful and something defines an "isb" macro in the wrong place (or if future code motion reverses the order of headers) that you'll silently lose your instruction barriers or otherwise miscompile your project.

Seems like it's more robust to do this with conditionally included headers and ARCH_HAS_XXX kconfigery, along the lines of how atomic.h works.

Reply:
sounds a good idea.